### PR TITLE
Python 3

### PR DESCRIPTION
**Pull request by kurtismullins**
**IMPORTANT NOTES**

- This is currently a work-in-progress.
- Do not merge this into `master` until all CI tests are passing and we have moved the Python 2.x code into another branch.

-----------------------------------------

### Description of Changes

This Pull-Request encompasses changes required to transition Quay from Python 2.7 to Python 3.8. 

#### Changes:

- Run `2to3` to convert Python 2.7 compatible syntax to Python 3.8 compatible syntax.
- Run `black` targeting Python 3.8 for appropriate code formatting.
- Replaces `gpgme` with `gpg` wrapper Python package
- Replaces `resumablehashlib` with `rehash` package.
- Correct usage of unicode strings and bytes, conversion.
- Remove backport libraries which are no longer required.
- Fix CI scripts, Dockerfiles, and other non-Python files to be target Python 3.8
- Update packages as necessary to be compatible with Python 3.8
- Update the pypi dependencies to more modern version

#### Issue:

- Master Epic: https://issues.redhat.com/browse/PROJQUAY-68

#### Known issues

- buildman has not been converted to Python3 yet
- app registry not working

#### UNIT TESTING
- [X] auth
- [x] buildtrigger
- [ ] buildman/test
- [x] config_app
- [x] data/cache
- [x] data/logs_model
- [x] data/migrations
- [x] data/model/test
- [x] data/model/oci/test
- [x] data/registry_model
- [x] data/test
- [x] data/users
- [X] digest
- [x] endpoints/test
- [x] endpoints/api
- [x] endpoints/appr/test
- [x] endpoints/v2/
- [x] endpoints/verbs
- [x] image
- [x] notifications
- [x] oauth
- [x] storage
- [x] test
- [x] util/config
- [X] util/ipresolver
- [X] util/migrate
- [x] util/registry
- [X] util/secscan
- [X] util/security
- [X] util/test
- [X] util/tufmetadata 
- [x] workers

**BREAKING CHANGE**

- Quay is no longer compatible with Python < 3.x

---

## Reviewer Checklist

- [ ] It works!
- [ ] Comments provide sufficient explanations for the next contributor
- [ ] Tests cover changes and corner cases
- [ ] Follows Quay syntax patterns and format

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,14 +1,15 @@
-FROM centos:7
+FROM centos:8
+#FROM registry.access.redhat.com/ubi8:8.1
```
Comment (josephschorr) [inline Dockerfile:2]: Remove this

File: Dockerfile.cirun [hunk 2]
```diff
@@ -1,7 +1,15 @@
 FROM quay-ci-base
+
+ENV ENCRYPTED_ROBOT_TOKEN_MIGRATION_PHASE remove-old-fields
```
Comment (josephschorr) [inline Dockerfile.cirun:3]: Not needed for now.

Also, do we still need the cirun file?

File: auth/test/test_credentials.py [hunk 3]
```diff
@@ -156,7 +156,7 @@ def test_invalid_unicode_robot(app):
     result, kind = validate_credentials("devtable+somerobot", token)
     assert kind == CredentialKind.robot
     assert not result.auth_valid
-    msg = "Could not find robot with specified username"
+    msg = "Could not find robot with username: devtable+somerobot and supplied password."
```
Comment (josephschorr) [inline auth/test/test_credentials.py:159]: Why this change?

File: buildtrigger/test/test_githosthandler.py [hunk 4]
```diff
@@ -93,9 +93,9 @@ def test_list_build_source_namespaces():
     ]
 
     found = get_bitbucket_trigger().list_build_source_namespaces()
-    found.sort()
+    sorted(found, key=lambda d: sorted(d.items()))
```
Comment (josephschorr) [inline buildtrigger/test/test_githosthandler.py:96]: This isn't actually going to sort anything, since it *returns* the sorted value

File: buildtrigger/test/test_githubhandler.py [hunk 5]
```diff
@@ -129,7 +129,7 @@ def test_list_build_source_namespaces(github_trigger):
     ]
 
     found = github_trigger.list_build_source_namespaces()
-    found.sort()
+    sorted(found, key=lambda d: sorted(d.items()))
```
Comment (josephschorr) [inline buildtrigger/test/test_githubhandler.py:132]: Ditto

File: conf/init/v3_migration.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+from app import app
```
Comment (josephschorr) [inline conf/init/v3_migration.py:1]: This was removed in HEAD

File: config_app/config_endpoints/common.py [hunk 7]
```diff
@@ -23,7 +23,7 @@ def truthy_bool(param):
 TYPE_CONVERTER = {
     truthy_bool: "boolean",
     str: "string",
-    basestring: "string",
+    str: "string",
```
Comment (josephschorr) [inline config_app/config_endpoints/common.py:26]: No longer needed

File: config_app/config_util/ssl.py [hunk 8]
```diff
@@ -28,7 +28,7 @@ def load_certificate(cert_contents):
         cert = OpenSSL.crypto.load_certificate(OpenSSL.crypto.FILETYPE_PEM, cert_contents)
         return SSLCertificate(cert)
     except OpenSSL.crypto.Error as ex:
-        raise CertInvalidException(ex.message[0][2])
+        raise CertInvalidException(str(ex))
```
Comment (josephschorr) [inline config_app/config_util/ssl.py:31]: If I recall, this broke the messages being displayed

File: data/database.py [hunk 9]
```diff
@@ -1398,7 +1398,7 @@ class BlobUpload(BaseModel):
     repository = ForeignKeyField(Repository)
     uuid = CharField(index=True, unique=True)
     byte_count = BigIntegerField(default=0)
-    sha_state = ResumableSHA256Field(null=True, default=resumablehashlib.sha256)
+    sha_state = ResumableSHA256Field(null=True, default=rehash.sha256)
```
Comment (josephschorr) [inline data/database.py:1401]: Are we 100% certain this is backwards compatible? Please add a TODO and JIRA to verify

File: data/fields.py [hunk 10]
```diff
@@ -17,42 +19,44 @@ def random_string(length=16):
...
             return None
 
-        sha_state = value.state()
-
-        # One of the fields is a byte string, let's base64 encode it to make sure
-        # we can store and fetch it regardless of default collocation.
-        sha_state[3] = base64.b64encode(sha_state[3])
-
-        return json.dumps(sha_state)
+        serialized_state = base64.b64encode(pickle.dumps(value)).decode("ascii")
```
Comment (josephschorr) [inline data/fields.py:38]: Please link this line in the JIRA to file: pickling will likely not work long-term, and there are back-compat concerns

File: data/fields.py [hunk 11]
```diff
@@ -180,27 +183,6 @@ def python_value(self, value):
 
             return LazyEncryptedValue(value, self)
 
-        def __eq__(self, _):
```
Comment (josephschorr) [inline data/fields.py:183]: Why were these removed?

File: data/logs_model/test/fake_elasticsearch.py [hunk 12]
```diff
@@ -75,7 +75,7 @@ def post_doc(url, request):
     def index_delete(url, request):
         index_name_or_pattern = url.path[1:]
         to_delete = []
-        for index_name in docs.keys():
+        for index_name in list(docs.keys()):
```
Comment (josephschorr) [inline data/logs_model/test/fake_elasticsearch.py:78]: Please file a JIRA for us to revert all these `list(somedict.keys())` changes; they aren't necessary unless we're accessing the list component of `.keys`

File: digest/checksums.py [hunk 13]
```diff
@@ -64,7 +64,7 @@ def compute_tarsum(fp, json_data):
             hashes.append(h)
         hashes.sort()
     except tarfile.ReadError as e:
-        if e.message != "empty file":
+        if str(e) != "empty file":
```
Comment (josephschorr) [inline digest/checksums.py:67]: This may not work; we need to test this. Please add a TODO

File: endpoints/api/discovery.py [hunk 14]
```diff
@@ -34,7 +34,7 @@
 TYPE_CONVERTER = {
     truthy_bool: "boolean",
     str: "string",
-    basestring: "string",
+    str: "string",
```
Comment (josephschorr) [inline endpoints/api/discovery.py:37]: Not needed anymore

File: Dockerfile (hunk 1 above)
Comment (kleesc) [reply Dockerfile]: Done

File: Dockerfile.cirun (hunk 2 above)
Comment (kleesc) [reply Dockerfile.cirun]: Don't think so with tox. Removed the file.

File: buildtrigger/test/test_githosthandler.py (hunk 4 above)
Comment (kleesc) [reply buildtrigger/test/test_githosthandler.py]: Fixed.

File: buildtrigger/test/test_githubhandler.py (hunk 5 above)
Comment (kleesc) [reply buildtrigger/test/test_githubhandler.py:132]: Fixed.

File: conf/init/v3_migration.py (hunk 6 above)
Comment (kleesc) [reply conf/init/v3_migration.py]: Removed the file.

File: config_app/config_endpoints/common.py (hunk 7 above)
Comment (kleesc) [reply config_app/config_endpoints/common.py:26]: Removed the duplicate `str: "string"`

File: data/database.py (hunk 9 above)
Comment (kleesc) [reply data/database.py:1401]: https://issues.redhat.com/browse/PROJQUAY-771

File: data/fields.py (hunk 10 above)
Comment (kleesc) [reply data/fields.py:38]: https://issues.redhat.com/browse/PROJQUAY-771

File: data/fields.py (hunk 11 above)
Comment (kleesc) [reply data/fields.py:183]: My understanding is that fields need to be hashable. 
In Python3, defining `__eq__` but not `__hash__`  will implicitly set `__hash__` to `None` and raise a TypeError when something calls hash on that object.
I think it will inherit that of its parent in Python2:
https://docs.python.org/3/reference/datamodel.html#object.__hash__
https://docs.python.org/2/reference/datamodel.html#object.__hash__

Looking at peewee. __hash__ was implemented on the base `Field` class: https://github.com/coleifer/peewee/blob/master/peewee.py#L4470
@kurtismullins Instead of removing the magic methods, can we just implement __hash__ calling the field class's __hash__?

File: data/logs_model/test/fake_elasticsearch.py (hunk 12 above)
Comment (kleesc) [reply data/logs_model/test/fake_elasticsearch.py:78]: https://issues.redhat.com/browse/PROJQUAY-771

File: endpoints/api/discovery.py (hunk 14 above)
Comment (kleesc) [reply endpoints/api/discovery.py]: Removed the duplicate.

File: digest/checksums.py (hunk 13 above)
Comment (kleesc) [reply digest/checksums.py:68]: Added TODO to add an empty tarfile test.

File: auth/test/test_credentials.py (hunk 3 above)
Comment (kleesc) [reply auth/test/test_credentials.py:159]: https://github.com/quay/quay/blob/master/data/model/user.py#L429
We were previously ignoring non-ascii characters, in the token, which is why the `UnicodeEncodeError` block was skipped (the non-ascii quotes were being stripped from the byte string)

https://github.com/quay/quay/blob/python3/data/model/user.py#L431
Since we're dealing with unicode now, we just try to encode the token as ascii to ensure that it's just ascii characters, and (as expected) get the exception.

File: data/fields.py (hunk 11 above)
Comment (kleesc) [reply data/fields.py:183]: @josephschorr Reverted the change and implemented the `__hash__` magic method instead:
```
def __hash__(self):
  return field_class.__hash__(self)
```

**Comment by kleesc** [comment]
@kurtismullins @thomasmckay @josephschorr @alecmerdler Ping

File: data/fields.py (hunk 11 above)
Comment (josephschorr) [reply data/fields.py:183]: Cool. Thanks

File: auth/test/test_credentials.py (hunk 3 above)
Comment (josephschorr) [reply auth/test/test_credentials.py:159]: Cool

File: storage/test/test_swift.py [hunk 15]
```diff
@@ -1,14 +1,16 @@
...
 
 from collections import defaultdict
 from mock import MagicMock, patch
 
+from swiftclient.client import ClientException, ReadableToIterable
+
 from storage import StorageContext
 from storage.swift import SwiftStorage, _EMPTY_SEGMENTS_KEY
-from swiftclient.client import ClientException
+from util.registry.generatorfile import GeneratorFile
```
[the remark this answers is not in the document]
Comment (kleesc) [reply storage/test/test_swift.py]: Removed the unused import.

The retry count is for https://github.com/quay/quay/blob/python3/storage/swift.py#L412
Apparently, `None > 0` was allowed in Python2.

File: storage/test/test_cloud_storage.py [hunk 16]
```diff
@@ -1,13 +1,14 @@
...
-from StringIO import StringIO
+from io import BytesIO
 
 import pytest
 
 import moto
 import boto
 
 from moto import mock_s3_deprecated as mock_s3
+from moto.s3 import s3_backend
```
[the remark this answers is not in the document]
Comment (kleesc) [reply storage/test/test_cloud_storage.py]: Removed

File: storage/test/#test_swift.py# [hunk 17]
```diff
@@ -0,0 +1,342 @@
+import _pyio as io
```
[the remark this answers is not in the document]
Comment (kleesc) [reply storage/test/#test_swift.py#]: Removed

File: endpoints/appr/registry.py [hunk 18]
```diff
@@ -197,6 +198,7 @@ def pull(namespace, package_name, release, media_type):
         metadata={"release": release, "mediatype": media_type},
     )
     json_format = request.args.get("format", None) == "json"
+    print("DATA:", data)
```
[the remark this answers is not in the document]
Comment (kleesc) [reply endpoints/appr/registry.py]: Removed

File: data/registry_model/registry_pre_oci_model.py [hunk 19]
```diff
@@ -0,0 +1,830 @@
+# pylint: disable=protected-access
```
[the remark this answers is not in the document]
Comment (kleesc) [reply data/registry_model/registry_pre_oci_model.py]: Removed
